include/boost/geometry/algorithms/detail/overlay/graph/traverse_graph.hpp

```diff
@@ -178,73 +178,89 @@ struct traverse_graph
     bool continue_traverse(Ring& ring,
             signed_size_type component_id,
             signed_size_type start_node_id,
-            signed_size_type current_node_id)
+            signed_size_type target_node_id)
     {
-        auto const current_turn_indices = get_turn_indices_by_node_id(m_turns, m_clusters,
-                current_node_id, allow_closed);
-
-        // Any valid node should always deliver at least one turn
-        BOOST_ASSERT(! current_turn_indices.empty());
+        signed_size_type current_node_id = target_node_id;
 
-        auto const next_target_nodes = get_target_nodes<target_operation>(m_turns, m_clusters,
-                current_turn_indices, component_id);
+        std::size_t iteration_count = 0;
 
-        if (next_target_nodes.empty())
+        // Keep traversing until it finds the start (successful finish), or it is stuck,
+        // or it find an already visited node during traversal.
+        // The iteration count is a defensive check to prevent endless loops and not iterate
+        // more than times there are turns (this should not happen).
+        while (iteration_count < m_turns.size())
```

Comment:
Best reviewing it with **Hide whitespace**.
Then you can see it just changes the recursive behaviour to a loop.